libbindgen/tests/tests.rs

```diff
@@ -0,0 +1,132 @@
+        Ok(bindings) => bindings.to_string(),
+        Err(_) => "".to_string(),
+    };
+
+    let mut buffer = String::new();
+    let f = try!(fs::File::open(&expected));
+    let _ = try!(BufReader::new(f).read_to_string(&mut buffer));
+
+    if output == buffer {
+        return Ok(());
+    }
+
+    println!("diff expected generated\n--- expected: {:?}\n+++ generated from: {:?}",
+             expected,
+             header);
+    for diff in diff::lines(&buffer, &output) {
+        match diff {
+            diff::Result::Left(l) => println!("-{}", l),
+            diff::Result::Both(l, _) => println!(" {}", l),
+            diff::Result::Right(r) => println!("+{}", r),
+        }
+    }
+    Err(Error::new(ErrorKind::Other, "Header and binding differ!"))
+}
+
+fn create_bindgen_builder(header: &PathBuf)
+                          -> Result<libbindgen::Builder, Error> {
+    let source = try!(fs::File::open(header));
+    let reader = BufReader::new(source);
+
+    // Scoop up bindgen-flags from test header
+    let line: String = try!(reader.lines().take(1).collect());
+    let flags =
+        match line.split("bindgen-flags:").last().and_then(shlex::split) {
+            Some(args) => args,
+            _ => Vec::with_capacity(2),
+        };
+
+    // Fool builder_from_flags() into believing it has real env::args_os...
+    // - add "bindgen" as executable name 0th element
+    // - add header filename as 1st element
+    // - prepend raw lines so they're in the right order for expected output
+    // - append the test header's bindgen flags
+    let args = vec![
+        "bindgen",
+        header.to_str().unwrap_or(""),
+        "--raw-line", "",
+        "--raw-line", "#![allow(non_snake_case)]",
+        "--raw-line", "",
+    ]
+        .into_iter()
+        .map(|s| s.to_string())
+        .chain(flags.into_iter());
+
+    builder_from_flags(args).map(|(builder, _)| builder.no_unstable_rust())
+}
+
+#[test]
+fn run_bindgen_tests() {
+    log::set_logger(|max_log_level| {
+            use env_logger::Logger;
+            let env_logger = Logger::new();
+            max_log_level.set(env_logger.filter());
+            Box::new(env_logger)
+        })
+        .expect("Failed to set logger.");
+
+    let manifest_env = env::var("CARGO_MANIFEST_DIR")
+        .expect("CARGO_MANIFEST_DIR not set!");
+    let manifest_dir = Path::new(&manifest_env);
+    let headers_dir = manifest_dir.join("tests").join("headers");
+
+    let entries = fs::read_dir(&headers_dir)
+        .expect("Couldn't read headers dir")
+        .map(|result| result.expect("Couldn't read header file"));
+
+    let headers: Vec<_> = entries.filter_map(|entry| {
+            match entry.path().extension().and_then(|s| s.to_str()) {
+                Some("h") | Some("hpp") => Some(entry.path()),
+                _ => None,
+            }
+        })
+        .collect();
+
+    let failures: Vec<_> = headers.iter()
+        .filter_map(|header| {
```

Comment:
I had a quick go at that when we were still shelling out to `rustc`, but couldn't figure out the rayon equivalent of `collect()`. I tried a few variations of `collect_into()` but… no. But hey, the whole test suite is so much faster now we can take our time. 😄